source/extensions/filters/http/istio_stats/istio_stats.cc

```diff
@@ -0,0 +1,669 @@
+      case Reporter::ServerSidecar: {
+        const auto ssl_info = info.downstreamAddressProvider().sslConnection();
+        const auto mtls =
+            ssl_info != nullptr && ssl_info->peerCertificatePresented();
+        tags_.push_back({context_.connection_security_policy_,
+                         mtls ? context_.mutual_tls_ : context_.none_});
+        break;
+      }
+      default:
+        tags_.push_back(
+            {context_.connection_security_policy_, context_.unknown_});
+        break;
+    }
+  }
+
+  // Peer metadata is populated after encode/decodeHeaders by MX HTTP filter,
+  // and after initial bytes read/written by MX TCP filter.
+  void populatePeerInfo(const StreamInfo::StreamInfo& info,
+                        const StreamInfo::FilterState& filter_state) {
+    const auto& filter_state_key =
+        config_->reporter() == Reporter::ServerSidecar ? "wasm.downstream_peer"
+                                                       : "wasm.upstream_peer";
+    const auto* object = filter_state.getDataReadOnly<
+        Envoy::Extensions::Filters::Common::Expr::CelState>(filter_state_key);
+    absl::optional<Istio::Common::WorkloadMetadataObject> peer;
+    if (object) {
+      const auto& node =
+          *flatbuffers::GetRoot<Wasm::Common::FlatNode>(object->value().data());
+      peer.emplace(Istio::Common::convertFlatNodeToWorkloadMetadata(node));
+    }
+    // Compute destination service with fallbacks.
+    absl::string_view service_host;
+    absl::string_view service_host_name;
+    const auto cluster_info = info.upstreamClusterInfo();
+    if (cluster_info && cluster_info.value()) {
+      const auto& filter_metadata =
+          cluster_info.value()->metadata().filter_metadata();
+      const auto& it = filter_metadata.find("istio");
+      if (it != filter_metadata.end()) {
+        const auto& services_it = it->second.fields().find("services");
+        if (services_it != it->second.fields().end()) {
+          const auto& services = services_it->second.list_value();
+          if (services.values_size() > 0) {
+            const auto& service = services.values(0).struct_value().fields();
+            const auto& host_it = service.find("host");
+            if (host_it != service.end()) {
+              service_host = host_it->second.string_value();
+              service_host_name =
+                  service_host.substr(0, service_host.find_first_of('.'));
+            }
+          }
+        }
+      }
+    }
+    if (service_host.empty() && !config_->disable_host_header_fallback_) {
+      const auto* headers = info.getRequestHeaders();
+      if (headers && headers->Host()) {
+        service_host = headers->Host()->value().getStringView();
+        service_host_name = service_host;
+      }
+    }
+    switch (config_->reporter()) {
+      case Reporter::ServerSidecar: {
+        tags_.push_back({context_.source_workload_,
+                         peer ? config_->resolve(peer->workload_name_)
```

Comment:
could there maybe be some opportunity optimizing this not to use a map lookup?

for example, can't we initialize these into variables when the connection is first established? so then we can directly access the variable and not do map operations

Reply:
Yes, I think that's the only way to further optimize this, but it requires some architectural changes to make lifecycle of metadata more accurate. There are two options:
* require TCP metadata for HTTP traffic, and then move the peer population to the filter constructor;
* require CONNECT tunneling (from ambient), which is equivalent to the first with a different protocol.

Without knowing a-priori that connections cannot be shared between peers, we can't do much but expect every request to have different per-request data.

Reply:
https://github.com/istio/istio/issues/36420

I think the performance is going to be better with native rewrite.

Reply:
Hmm yes, it would be great to rewrite TCP MX in native and use that instead of HTTP where possible, i'm sure there would be CPU usage benefits as well as latency improvements.